### PR TITLE
typespecs on some result functions too tight

### PR DESCRIPTION
**Pull request by jvantuyl**
A number of functions take an `on_ok` or `on_error` parameter.  This is either an arbitrary response value or a function which returns one.  In many cases, this is typed as a result-tuple.

This commit loosens those typespecs from `t` to `any`.

### CODE REVIEW

**Comment by yordis** [comment]
Hey there, thank you so much for contributing.

Could you share some code samples that such typespec is causing trouble?

Based on the code, the inputs match against the ok-error tuple, and the outputs are wrapped.

So I am wondering what piece of code is causing trouble per se.

**Comment by jvantuyl** [comment]
TL;DR:  Don't merge this.  I eliminated a Dialyzer warning but it was on accident, not because of these changes.

For posterity, I'll go ahead and explain what was going on...

When this first popped up, we were using version 0.3.1.  We have something like this (redacted heavily because proprietary code):

```elixir
defmodule Redacted do
  def handle(_parent, args, resolution) do
    entity = Graphql.current_entity(resolution)
    some_transfer_id = OurCompany.id().new()

    with true <- is_something_account_owner?(entity, args.something_account_id),
         :ok <- check_something_account_ownership(entity, args),
         {:ok, something_plan_transfer_data} <-
           get_something_plan_transfer_data(entity, args[:something_plan_transfer_data], args.type),
         :ok <-
           OurCompany.command_dispatcher().dispatch(%InitiateTransfer{
             ...
           }) do
      {:ok, %{something_transfer_id: something_transfer_id}}
    else
      false -> {:error, :bad_request}
      error -> error
    end

  defp check_something_account_ownership(entity, args) do
    from(
      sa in SomethingAccount,
      where:
        sa.id == ^args.something_account_id and sa.owner_id == ^entity.id and is_nil(sa.archived_at) and
          sa.ownership_validation_status == :VALID
    )
    |> Repo.one()
    |> OnePiece.Result.reject_nil(:something_account_failed)
    |> OnePiece.Result.when_ok(:ok)
  end
end
```

The issue was the line:  `:ok <- check_something_account_ownership(entity, args)`

Because `check_something_account_ownership/2` effectively returns `OnePiece.Result.when_ok(:ok)`, it gets the typespec for that function.  Namely: `@spec when_ok(result :: t, on_ok :: (any -> any) | any) :: t`

When `on_ok` is a value, the return type is *not* a result-tuple (and, thus, not of type `t`).  That gave me this error:
```
lib/our_company/web/graphql/resolver/mutations/redacted.ex:16:pattern_match
The pattern can never match the type.

Pattern:
:ok

Type:
{:error, _} | {:ok, _}
```

On closer inspection, I think the actual issue I was having was fixed with 0.4.0 (specifically in #37).  We hadn't updated for reasons (tests were breaking and never had time to investigate).  In doing this patch, I got the tests passing and assumed that my changes had eliminated the above error.  But now I think I was wrong there.  It appears that basing my changes off of 0.4.0 is what actually fixed the error.

Now that I've dug a bit deeper, I think my changes aren't correct because of some subtly different behavior of `on_ok` between different helper functions.  Specifically, in `when_ok/2`, the return values aren't explicitly wrapped in `ok`, but in `map_ok/2` they are, but in `map_ok_or/2` they aren't, etc.  I had assumed these should behave the same, but I guess they don't.  So these "fixes" were to align functions that aren't intended to be aligned.

Oops... yeah... I'll go ahead and close this.  That said, it's pretty easy to miss that distinction.  I see the note in the docs, but the meaning of `on_ok` changing in `map_ok/2` versus `when_ok/2` is a bit confusing.

**Comment by yordis** [comment]
> (redacted heavily because proprietary code)

Don't worry. I can repeat back to you almost every line 😄 you could even remove more code and I can put it back together in a few seconds 😄 

---

> It's not clear to me why the functions are named what they are.

The naming comes from Rust Result and other packages alike. 

The idea behind it is as follows:

1. Pipeline continues the `ok-tuple` response pipeline (all the map_*)
2. Pipeline `could` exists/bail out the `ok-tuple` response pipeline (all the when_*)

Both functions intend to "remove" the control flow **from your function definitions** as much as you can. At the same time, we have some control over the pipelining.

In an ideal scenario, your use-case entry-point function does most of the control flow.


---

> We hadn't updated for reasons (tests were breaking and never had time to investigate)

That is no bueno at all, definitely wouldn't expect such an upgrade to be a "breaking change", did I mess it up anyhow? Please share the learnings if so, a positive feedback loop is the only way to improve the packages in the ecosystem.

**Comment by jvantuyl** [comment]
## Result Types

Gotcha on the naming and "removing control flow from the functions themselves".  I respect that goal, though I find it easier to pipe into case statements for flow control.  It might just be that I'm bad at memorizing a bunch of similarly named helper functions.

I tend to write pipelines like:
```elixir
def some_func() do
  do_a_thing!()
  |> and_another_thing()
  |> and_yet_another_thing()
  |> and_now_something_that_returns_a_result_tuple()
  |> case do
    {:ok, value} ->
      value
    {:error, _reason} = err ->
      throw(err)
  end
  |> do_next_thing!()
  |> then_something_with_result_tuples_again()
  # and no case here because we want to return a result tuple
catch
  {:error, _reason} = err ->
    err
end
```

It's not everyone's cup-of-tea, but I find that there's rarely confusion about what's going on.  You get control-flow specified in the pipeline and you get early-bailout without having to weave the errors all the way through.  The `catch` bail-out approach also works well in the presence of recursion or even just deep call stacks.

Conversely, if I was to take a survey of the devs I've worked with, I'd be willing to bet that most of them couldn't reliably identify what was going on in some of those pipelines.  `when_ok` is pretty evident.  `map_ok`, `tap_ok`, `map_ok_or`, and friends get a bit more inscrutable.  And with function parameters, they get pretty obtuse.

That is, of course, just my opinion.  And it's not really one that makes sense for `one_piece_result` since that style is kind of the point.  I only really got here because, at first, I incorrectly thought the typespecs weren't right.

## Upgrading Issues

As for the not upgrading, it's sort of the same.  We were running `one_piece_commanded` version 0.15.2.  When upgrading all the way, stuff exploded in ways that I still haven't gotten to the bottom of.  Part of why we were having trouble figuring out what was busted was a complete lack of attention to Dialyzer.  It just kind of languished under the combined weight of the depth of the macros, the breakage from some of the changes in 0.19.0, and a widespread assumption that we shouldn't upgrade the `one_piece_*` components separately from each other.

I recently set out to at least fix the typespecs.  That is part of why I ended up noticing the typespec issue you fixed in #37.  We were still encountering it because we hadn't upgraded.  I briefly thought the other typespecs were also off until I finally figured out what was generating the other errors on our end.  This ill-fated PR was just a step in the process. 😅

As for what we were doing wrong.  We broke some patterns built into `one_piece_commanded`, that was feeding Dialyzer unexpectedly shredded typespecs, and then we lied to Dialyzer a lot on top of that.  Some of the typespecs that had grown to "make the errors go away" in the absence of understanding were... simultaneously a bit inspired and a bit deranged.

Eventually I tracked down most of the errors to a bunch of `new` functions we defined and sparingly use.  They don't conform to the typespec inherited from the version they override.  This resulted in errors stemming from invisible code (due to the macro) and just a bunch of things that came together in non-obvious ways.

Consider the following:

```elixir
defmodule SomeError do
  use OnePiece.Result.ValueObject

  @type t() :: %__MODULE__{...}

  # other stuff

  def new(error_message) when is_binary(error_message) do
    %__MODULE__{message: error_message}
  end
end
```

This gives a Dialyzer error because the typespec from the macro doesn't accept a binary and states that it returns a result-tuple (which this doesn't).

But what if we add this?
```elixir
@spec new(String.t()) :: t()
```

Now it passes Dialyzer because the specs magically combine.

But what if our exciting constructor takes another struct like this?
```elixir
def new(%OtherErrorStruct{} = error) do
  %__MODULE__{inner_error: error}
end
```

This fails, but not because of the struct!  Since it's trivially a map, we only get a failure on the return type.

So now what if add a typespec to that like this?
```elixir
@spec new(OtherErrorStruct.t()) :: t()
```

This gets even weirder.  Rather than invisibly building a passing (but very busted) typespec, it fails with errors about overlapping typespecs.  Because map and structs technically overlap.

It gets clearer when you expand the macros, because you end up seeing sequences like this:
```elixir
@spec new(map()) :: {:ok, %__MODULE__{}}
def new(attrs) do
  ...
  {:ok, value}
end
defoverridable new: 1
@type t() :: %__MODULE__{}
@spec new(binary()) :: t()
def new(msg) when is_binary(msg) do
  ...
   value
end
```

It's a story as old as time.  "Macro does something; It's invisible; We end up compiling nonsense; But Dialyzer doesn't speak human enough for us to understand."  We had about 20 of these failures and another 10 silently passing but not really typed correctly.  Everybody had been ignoring Dialyzer because nobody understood the errors.

Everybody knew something was busted but nobody wanted to touch it because they couldn't get to the bottom of it.  So... yeah... I showed up with an overwhelming OCD urge to placate Dialyzer and a slightly masochistic streak.  And here we are.

Still trying to get all of those fixed, too.  Ironically, we don't use half of the `new` functions.  But the presence of the "passing but incorrect" ones make this a long process to audit them all.  Personally, I find this to be a sort of hole in `defoverridable`.

There really needs to be a way to override the typespec, too.  Maybe something like `@reset_spec new/1`.  Or maybe just recognize the differing `Env` for the macro and shadow it completely if anything gets defined at all in the code that uses it.  Or make it an error to change the type-signature of a function when overriding it.

I don't really have a solution, but at least this wall of text will be there in Google for the next time somebody accidentally creates this sort of problem.
